# Topic-Wordnet and Word Sense Disambiguation| NLP Assignment-2| Group-111

### Group Members Name with Student ID and Contribution:
### Student 1: AHIRE PAWAN ANIL (2023AC05618) - 100%
### Student 2: TUSHAR SHANDILYA (2023AC05579) - 100%
### Student 3: SAURABH JALENDRA (2023AC05912) - 100%
### Student 4: MONICA MALIK (2023AC05875) - 100%
### Student 5: KOLEKAR SMITA RAJESH (2023AC05092) - 100%

## 1. a)DDownload the datase  b)Perform pre-processing steps  c)Normalize review column by using Stemming or Lemmatization  d)Demonstrate use of Sentiwordnet to calculate the senti_scoret

In [10]:
import pandas as pd

In [11]:
# Load the dataset
MOVIE = pd.read_csv("Movie.csv")

#### The dataset is loaded into a DataFrame named MOVIE. 

### Justification
The dataset contains two columns: text (movie reviews) and label (binary labels, likely 0 for negative and 1 for positive sentiment).

The dataset is successfully loaded, and the DataFrame is ready for further processing.


In [12]:
# Check the first few rows
print(MOVIE.head(10))

                                                text  label
0  I grew up (b. 1965) watching and loving the Th...      0
1  When I put this movie in my DVD player, and sa...      0
2  Why do people who do not know what a particula...      0
3  Even though I have great interest in Biblical ...      0
4  Im a die hard Dads Army fan and nothing will e...      1
5  A terrible movie as everyone has said. What ma...      0
6  Finally watched this shocking movie last night...      1
7  I caught this film on AZN on cable. It sounded...      0
8  It may be the remake of 1987 Autumn's Tale aft...      1
9  My Super Ex Girlfriend turned out to be a plea...      1


#### Displays the first 10 rows of the dataset. Each row contains a movie review and its corresponding label.

### Justification
The head() function displays the first 5 rows of the dataset.

The text column contains the movie reviews, and the label column contains the sentiment labels (0 or 1).

This output confirms that the dataset is structured correctly and contains the expected data.

In [13]:
# Get dataset information
print(MOVIE.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40000 entries, 0 to 39999
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   text    40000 non-null  object
 1   label   40000 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 625.1+ KB
None


#### Provides information about the dataset, including the number of entries (40,000), the columns (text and label), and their data types. It also shows that there are no missing values.

### Justification
This output confirms that the dataset is clean and ready for preprocessing.

In [14]:
# Get statistical summary
print(MOVIE.describe(include='all'))

                                                     text         label
count                                               40000  40000.000000
unique                                              39723           NaN
top     Hilarious, clean, light-hearted, and quote-wor...           NaN
freq                                                    4           NaN
mean                                                  NaN      0.499525
std                                                   NaN      0.500006
min                                                   NaN      0.000000
25%                                                   NaN      0.000000
50%                                                   NaN      0.000000
75%                                                   NaN      1.000000
max                                                   NaN      1.000000


#### Provides a statistical summary of the dataset. For the text column, it shows the count, unique values, most frequent text, and frequency of the most common text. For the label column, it shows the mean, standard deviation, min, max, and quartiles.

### Justification
#### For the text column:

Count: 40,000 non-null entries.

Unique: 39,723 unique reviews, meaning some reviews are repeated.

Top: The most frequent review is shown, along with its frequency (4  im### es).

For the label column:

Mean: 0.4995, indicating that the dataset is almost balanced (close to 50% positive and 50% negative reviews).

Std: 0.500006, showing the standard deviation of the labels.

Min/Max: Labels range from 0 to 1.

### 1b. Preprocessing: Remove Punctuations, Numbers, Special Characters, and Stop Words

In [15]:
import re
import nltk
from nltk.corpus import stopwords

# Download stopwords
nltk.download('stopwords')
stop_words = set(stopwords.words('english'))

def preprocess_text(text):
    # Remove punctuations, numbers, and special characters
    text = re.sub(r'[^\w\s]', '', text)
    text = re.sub(r'\d+', '', text)
    
    # Convert to lowercase
    text = text.lower()
    
    # Remove stopwords
    text = ' '.join([word for word in text.split() if word not in stop_words])
    
    return text

# Apply preprocessing to the 'text' column
MOVIE['cleaned_text'] = MOVIE['text'].apply(preprocess_text)

# Check the cleaned text
print(MOVIE[['text', 'cleaned_text']].head(10))


[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\tushandilya\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


                                                text  \
0  I grew up (b. 1965) watching and loving the Th...   
1  When I put this movie in my DVD player, and sa...   
2  Why do people who do not know what a particula...   
3  Even though I have great interest in Biblical ...   
4  Im a die hard Dads Army fan and nothing will e...   
5  A terrible movie as everyone has said. What ma...   
6  Finally watched this shocking movie last night...   
7  I caught this film on AZN on cable. It sounded...   
8  It may be the remake of 1987 Autumn's Tale aft...   
9  My Super Ex Girlfriend turned out to be a plea...   

                                        cleaned_text  
0  grew b watching loving thunderbirds mates scho...  
1  put movie dvd player sat coke chips expectatio...  
2  people know particular time past like feel nee...  
3  even though great interest biblical movies bor...  
4  im die hard dads army fan nothing ever change ...  
5  terrible movie everyone said made laugh cameo ... 

#### The cleaned_text column is created, which contains the preprocessed text after a)Removing punctuations, numbers, and special characters. b)Removing stopwords.The output shows the original text and the cleaned text for the first 10 rows.

### Justification:

The preprocessing function:

Removes punctuations and numbers using regex.

Converts text to lowercase.

Removes stopwords (common words like "I," "the," "and" that do not contribute much to sentiment).

The output shows the original text and the cleaned text for the first 10 rows. The cleaned text is free of noise and ready for further analysis.

### 1c. Normalize the review column using Stemming or Lemmatization

In [16]:
from nltk.stem import WordNetLemmatizer

# Download WordNet
nltk.download('wordnet')
lemmatizer = WordNetLemmatizer()

def lemmatize_text(text):
    return ' '.join([lemmatizer.lemmatize(word) for word in text.split()])

# Apply lemmatization to the 'cleaned_text' column
MOVIE['lemmatized_text'] = MOVIE['cleaned_text'].apply(lemmatize_text)

# Check the lemmatized text
print(MOVIE[['cleaned_text', 'lemmatized_text']].head(10))

[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\tushandilya\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


                                        cleaned_text  \
0  grew b watching loving thunderbirds mates scho...   
1  put movie dvd player sat coke chips expectatio...   
2  people know particular time past like feel nee...   
3  even though great interest biblical movies bor...   
4  im die hard dads army fan nothing ever change ...   
5  terrible movie everyone said made laugh cameo ...   
6  finally watched shocking movie last night dist...   
7  caught film azn cable sounded like would good ...   
8  may remake autumns tale eleven years director ...   
9  super ex girlfriend turned pleasant surprise r...   

                                     lemmatized_text  
0  grew b watching loving thunderbird mate school...  
1  put movie dvd player sat coke chip expectation...  
2  people know particular time past like feel nee...  
3  even though great interest biblical movie bore...  
4  im die hard dad army fan nothing ever change g...  
5  terrible movie everyone said made laugh cameo ... 

#### We have used Lemmitization for the normalization process. The lemmatized_text column is created, which contains the lemmatized version of the cleaned text. The output shows the cleaned text and the lemmatized text for the first 10 rows.

### Justification:

Lemmatization reduces words to their base or root form (e.g., "watching" → "watch," "movies" → "movie").

The output shows the cleaned text and the lemmatized text for the first 10 rows. Lemmatization ensures that different forms of the same word are treated as a single entity, improving the consistency of the text data.

### 1d. Use SentiWordNet to calculate the sentiment score of each review

In [17]:
from nltk.corpus import sentiwordnet as swn
from nltk.tokenize import word_tokenize

# Download necessary NLTK data
nltk.download('punkt')
nltk.download('sentiwordnet')

def get_sentiment_score(text):
    tokens = word_tokenize(text)
    pos_score = 0
    neg_score = 0
    for token in tokens:
        synsets = list(swn.senti_synsets(token))
        if synsets:
            synset = synsets[0]  # Use the first synset
            pos_score += synset.pos_score()
            neg_score += synset.neg_score()
    return pos_score - neg_score  # Net sentiment score

# Apply sentiment scoring to the 'lemmatized_text' column
MOVIE['senti_score'] = MOVIE['lemmatized_text'].apply(get_sentiment_score)

# Check the sentiment scores
print(MOVIE[['lemmatized_text', 'senti_score']].head(10))

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\tushandilya\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package sentiwordnet to
[nltk_data]     C:\Users\tushandilya\AppData\Roaming\nltk_data...
[nltk_data]   Package sentiwordnet is already up-to-date!


                                     lemmatized_text  senti_score
0  grew b watching loving thunderbird mate school...        1.875
1  put movie dvd player sat coke chip expectation...        1.750
2  people know particular time past like feel nee...        0.000
3  even though great interest biblical movie bore...       -0.125
4  im die hard dad army fan nothing ever change g...        1.625
5  terrible movie everyone said made laugh cameo ...       -1.375
6  finally watched shocking movie last night dist...        3.875
7  caught film azn cable sounded like would good ...        0.603
8  may remake autumn tale eleven year director ma...        2.375
9  super ex girlfriend turned pleasant surprise r...       10.875


#### The senti_score column is created, which contains the sentiment score for each review. The output shows the lemmatized text and the corresponding sentiment score for the first 10 rows.

### Justification:

SentiWordNet assigns positive (pos_score) and negative (neg_score) sentiment scores to words.

The net sentiment score is calculated as pos_score - neg_score.

The output shows the lemmatized text and the corresponding sentiment score for the first 5 rows. Higher scores indicate more positive sentiment, while lower or negative scores indicate more negative sentiment.

## 2. Take some sample reviews from the dataset and demonstrate the following: (5 Marks)

### a. Synsets

In [19]:
from nltk.corpus import wordnet

# Taking a random word Example from the review sample and displaying it's Synests
word = "disappointed"
synsets = wordnet.synsets(word)
print(f"Synsets for '{word}': {synsets}")

Synsets for 'disappointed': [Synset('disappoint.v.01'), Synset('defeated.s.02')]


#### Synsets are sets of synonyms in WordNet that represent different senses of a word.Each synset contains information about the word's meaning, part of speech, and usage.

### Justification:

Synsets are sets of synonyms that represent different senses of a word.

For "disappointed," there are two synsets:

disappoint.v.01: The verb sense, meaning to fail to meet expectations.

defeated.s.02: The adjective sense, meaning feeling let down or defeated.

This output demonstrates how WordNet captures multiple meanings of a word.

### b. Synonyms and Antonyms

In [20]:
# Taking a random word Example from the review sample and displaying it's Synonyms and Antonyms
word = "interest"
synsets = wordnet.synsets(word)
for synset in synsets:
    print(f"Synset: {synset.name()}")
    print(f"Synonyms: {[lemma.name() for lemma in synset.lemmas()]}")
    print(f"Antonyms: {[lemma.antonyms()[0].name() for lemma in synset.lemmas() if lemma.antonyms()]}")

Synset: interest.n.01
Synonyms: ['interest', 'involvement']
Antonyms: []
Synset: sake.n.01
Synonyms: ['sake', 'interest']
Antonyms: []
Synset: interest.n.03
Synonyms: ['interest', 'interestingness']
Antonyms: ['uninterestingness']
Synset: interest.n.04
Synonyms: ['interest']
Antonyms: []
Synset: interest.n.05
Synonyms: ['interest', 'stake']
Antonyms: []
Synset: interest.n.06
Synonyms: ['interest', 'interest_group']
Antonyms: []
Synset: pastime.n.01
Synonyms: ['pastime', 'interest', 'pursuit']
Antonyms: []
Synset: interest.v.01
Synonyms: ['interest']
Antonyms: ['bore']
Synset: concern.v.02
Synonyms: ['concern', 'interest', 'occupy', 'worry']
Antonyms: []
Synset: matter_to.v.01
Synonyms: ['matter_to', 'interest']
Antonyms: []


#### Synonyms: Lists words that have similar meanings in that context,Antonyms: Lists words with opposite meanings. Displays the synsets, synonyms, and antonyms for the word "difficult".

### Justification:

For each synset of "interest," the code retrieves synonyms and antonyms.

For example, in interest.n.03, the synonym is "interestingness," and the antonym is "uninterestingness."

This output demonstrates how WordNet captures relationships between words, including synonyms and antonyms.

### c. Hyponym and Hypernym

In [37]:
# Taking a random word Example from the movie review sample and displaying it's Hyponym and Hypernym
word = "people"
synsets = wordnet.synsets(word)
for synset in synsets:
    print(f"Synset: {synset.name()}")
    print(f"Hyponyms: {[lemma.name() for lemma in synset.hyponyms()]}")
    print(f"Hypernyms: {[lemma.name() for lemma in synset.hypernyms()]}")

Synset: people.n.01
Hyponyms: ['age_group.n.01', 'ancients.n.01', 'baffled.n.01', 'blind.n.01', 'blood.n.05', 'brave.n.02', 'business_people.n.01', 'chosen_people.n.01', 'class.n.03', 'clientele.n.01', 'coevals.n.01', 'damned.n.01', 'dead.n.01', 'deaf.n.01', 'defeated.n.01', 'disabled.n.01', 'doomed.n.01', 'enemy.n.03', 'episcopacy.n.01', 'folk.n.01', 'free.n.01', 'homebound.n.01', 'initiate.n.03', 'living.n.02', 'lobby.n.02', 'mentally_retarded.n.01', 'migration.n.02', 'nation.n.02', 'nationality.n.01', 'network_army.n.01', 'peanut_gallery.n.01', 'peoples.n.01', 'pocket.n.07', 'poor_people.n.01', 'populace.n.01', 'population.n.01', 'rank_and_file.n.02', 'retreated.n.01', 'rich_people.n.01', 'sick.n.01', 'smart_money.n.03', 'timid.n.01', 'tradespeople.n.01', 'unconfessed.n.01', 'unemployed_people.n.01', 'uninitiate.n.01', 'womankind.n.01', 'wounded.n.01']
Hypernyms: ['group.n.01']
Synset: citizenry.n.01
Hyponyms: ['achaean.n.02', 'aeolian.n.02', 'country_people.n.01', 'dorian.n.02', 'e

#### This demonstrates the hierarchical structure of WordNet, where words are organized into broader (hypernyms) and narrower (hyponyms) categories.Displays the hyponyms (more specific terms) and hypernyms (more general terms) for the word "people".

### Justification:

Hyponyms are more specific terms (e.g., "age_group" is a type of "people").

Hypernyms are more general terms (e.g., "people" is a type of "group").

This output demonstrates the hierarchical structure of WordNet, where words are organized into broader (hypernyms) and narrower (hyponyms) categories.



### d. WordNet Path Similarity

In [38]:
# Taking two random words Example from the movie review sample and displaying it's WordNet Path Similarity
word1 = "actor"
word2 = "director"
synset1 = wordnet.synsets(word1)[0]
synset2 = wordnet.synsets(word2)[0]
similarity = synset1.path_similarity(synset2)
print(f"Path similarity between '{word1}' and '{word2}': {similarity}")

Path similarity between 'actor' and 'director': 0.125


#### Displays the path similarity between the words "actor" and "director".Score of 0.125 between "actor" and "director" suggests that they are somewhat related but not very close in the WordNet hierarchy.

### Justification:

Path similarity measures how closely related two words are based on the shortest path between their synsets in the WordNet hierarchy.

A score of 0.125 indicates that "actor" and "director" are somewhat related (both are roles in the film industry) but not very close in the hierarchy.

This output demonstrates how WordNet quantifies the relatedness between words.

### e. Word Sense Disambiguation


In [1]:
from nltk.wsd import lesk

# Taking a random sentence Example from the movie review sample and displaying Word Sense Disambiguation of word 'Production'
sentence = "but the primitive production values of this film would cause any viewer to become bored"
word = "production"
sense = lesk(sentence.split(), word)
print(f"Disambiguated sense for '{word}': {sense}")


Disambiguated sense for 'production': Synset('production.n.08')


#### Word Sense Disambiguation (WSD): Determines the correct sense of a word in a given context.The Lesk algorithm is used to disambiguate the word "production" in the sentence.The output shows that the correct sense of "production" in this context is production.n.08, which refers to the process of creating something (e.g., a film).



### Justification:

The Lesk algorithm identifies the correct sense of a word in a given context.

For "production" in the sentence, the correct sense is production.n.08, which refers to the process of creating something (e.g., a film).

This output demonstrates how NLP tools can disambiguate word senses based on context.